### PR TITLE
[Model] Fix array out of bounds when rank > 2.

### PR DESCRIPTION
**Pull request by Duyi-Wang**

### CODE REVIEW

**Review by pujiang2018 — APPROVED** [review]
Next version, we may not need FP32 for logits.

**Comment by Duyi-Wang** [comment]
> Next version, we may not need FP32 for logits.
It can reduce communication overhead, but we still need to convert to FP16 when passed to Python for sampling....
We can do the convert and reorder at the same time.
